src/controllers/krate/versions.rs

```diff
@@ -21,27 +27,217 @@ pub async fn versions(state: AppState, Path(crate_name): Path<String>) -> AppRes
     })
     .await
 }
+
+fn list_by_date(
+    krate: &Crate,
+    options: Option<&PaginationOptions>,
+    req: &Parts,
+    conn: &mut PgConnection,
+) -> AppResult<PaginatedVersionsAndPublishers> {
+    let mut query = krate
+        .all_versions()
+        .left_outer_join(users::table)
+        .select((versions::all_columns, users::all_columns.nullable()));
+
+    if let Some(options) = options {
+        if let Page::Seek(ref seek) = options.page {
+            let (created_at, id) = seek.decode::<seek::Date>().map(|s| (s.0, s.1))?;
+            query = query.filter(
+                versions::created_at
+                    .eq(created_at)
+                    .and(versions::id.lt(id))
+                    .or(versions::created_at.lt(created_at)),
+            )
+        }
+        query = query.limit(options.per_page);
+    }
+
+    query = query.order((versions::created_at.desc(), versions::id.desc()));
+
+    let data: Vec<(Version, Option<User>)> = query.load(conn)?;
+    let mut next_page = None;
+    if let Some(options) = options {
+        next_page = next_seek_params(&data, options, |last| {
+            seek::Date(last.0.created_at, last.0.id)
+        })?
+        .map(|p| req.query_with_params(p));
+    };
+
+    // Since the total count is retrieved through an additional query, to maintain consistency
+    // with other pagination methods, we only make a count query while data is not empty.
+    let total = if !data.is_empty() {
+        versions::table.count().get_result(conn)?
+    } else {
+        0
+    };
+
+    Ok(PaginatedVersionsAndPublishers {
+        data,
+        meta: ResponseMeta { total, next_page },
+    })
+}
+
+// Unfortunately, Heroku Postgres has no support for the semver PG extension.
+// Therefore, we need to perform both sorting and pagination manually on the server.
+fn list_by_semver(
+    krate: &Crate,
+    options: Option<&PaginationOptions>,
+    req: &Parts,
+    conn: &mut PgConnection,
+) -> AppResult<PaginatedVersionsAndPublishers> {
+    let (data, total) = if let Some(options) = options {
+        // Sorting by semver but opted for id as the seek key because num can be quite lengthy,
+        // while id values are significantly smaller.
+        let mut sorted_versions = IndexMap::new();
+        for result in krate
+            .all_versions()
+            .select((versions::id, versions::num))
+            .load_iter::<(i32, String), DefaultLoadingMode>(conn)?
+        {
+            let (id, num) = result?;
+            sorted_versions.insert(id, (num, None));
+        }
+        sorted_versions.sort_by_cached_key(|_, (num, _)| Reverse(semver::Version::parse(num).ok()));
+
+        let mut idx = Some(0);
+        if let Page::Seek(ref seek) = options.page {
+            idx = seek
+                .decode::<i32>()
+                .map(|id| sorted_versions.get_index_of(&id))?
+                .filter(|i| i + 1 < sorted_versions.len())
+                .map(|i| i + 1);
+        }
```

Comment:
do we need an `else` branch here? I assume for `Page::Unspecified` doing nothing here is correct, but should we handle the `Page::Numeric` case in some way? the surrounding code should make this impossible, but technically `list_by_semver()` could be called with `Page::Numeric`.

Reply:
Great catch! You're right. Although we've disabled the option in the caller, the function itself doesn't technically prevent calling with the page-enabled option. However, we currently don't offer offset-base pagination capability and don't plan to in the future due to performance concerns. Perhaps we could add a NOTE comment clarifying this and simply return an not supported error. What do you think?

Reply:
I'd say `assert!()` + doc comment clarifying that the fn will panic in this case should probably be sufficient